lib/workers.js

```diff
@@ -1,213 +1,126 @@
+const genericPool = require('generic-pool');
 const { v4: uuidv4 } = require('uuid');
 const debug = require('debug')('prairielearn:' + path.basename(__filename, '.js'));
+const util = require('util');
 
 const logger = require('./logger');
 const config = require('./config');
 const load = require('./load');
 const codeCaller = require('./code-caller');
 
-/*
-At any time we have a set of pythonCallers, which are python workers
-that can process jobs, and callbacks, which are the jobs
-themselves. We might have more than enough pythonCallers, so there are
-some sitting around idle, or we might have not enough, so that we need
-to queue up callbacks until we get an available pythonCaller to handle
-them.
+/**
+ * This module maintains a pool of PythonCaller workers, which are used any
+ * time we need to execute Python code (elements, question code, etc.).
+ *
+ * Load is reported in:
+ * - python_worker_active: number of python workers currently processing jobs/callbacks
+ * - python_worker_idle: number of python workers available for incoming jobs/callbacks
+ * - python_callback_waiting: number of queued jobs/callbacks waiting for an available worker
+ */
 
-We maintain two FIFO queues, one for available pythonCallers (if any)
-and one for waiting callbacks (if any). At any time at least one of
-these queues must be empty.
-
-Load is reported in:
-- python_worker_active: number of python workers currently processing jobs/callbacks
-- python_worker_idle: number of python workers available for incoming jobs/callbacks
-- python_callback_waiting: number of queued jobs/callbacks waiting for an available worker
-
-*/
-
-let pythonCallers = []; // all pythonCallers, whether available for work or currently busy
-let availablePythonCallers = []; // FIFO queue for available pythonCallers waiting for a job/callback
-let waitingCallbacks = []; // FIFO queue for jobs/callbacks waiting for an available pythonCaller
+/** @type {import('generic-pool').Pool<codeCaller.PythonCaller>} */
+let pool = null;
 
 module.exports = {
   init() {
     debug('init()');
-    if (!config.useWorkers) {
-      return;
-    }
-    var numWorkers = config.workersCount;
-    if (numWorkers == null) {
-      numWorkers = Math.ceil(config.workersPerCpu * os.cpus().length);
-    }
-    for (let i = 0; i < numWorkers; i++) {
-      const pc = new codeCaller.PythonCaller();
-      pc.number = i;
-      pythonCallers.push(pc);
-      load.startJob('python_worker_idle', pc.uuid);
-      availablePythonCallers.push(pc);
-    }
-    module.exports._warmUpWorkers();
+    const numWorkers = config.workersCount ?? Math.ceil(config.workersPerCpu * os.cpus().length);
+    pool = genericPool.createPool(
+      {
+        create: async () => {
+          const pc = new codeCaller.PythonCaller();
+          pc.ensureChild();
+          load.startJob('python_worker_idle', pc.uuid);
+          return pc;
+        },
+        destroy: async (pc) => {
+          pc.done();
```

Comment:
Do we need a `load.endJob('python_worker_idle', pc.uuid);` here?

Reply:
Good call! Yes, I'll add that.

Reply:
Done in aa6ad64.